Multi-Z Support for Lazy Templates | Cleans up some turf flag misuse

ZephyrTFA opened the pull request:
## About The Pull Request

Adds multi-z support for lazy templates
Also fixes some improper use and placement for turf flags
## Why It's Good For The Game

Shadow needs/wants this for bit runner maps.
Turf flags are also why lava has been generating in places it shouldnt. (inside of ruins)
## Changelog
:cl:
fix: Lava can no longer occasionally generate inside of previously loaded templates and breach and/or destroy shit
/:cl:

ZephyrTFA committed:
some cleanup on var names and codedoc

ZephyrTFA committed:
yeah this wasnt fun

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -117,20 +117,59 @@
+	place_on_top = FALSE,
+	new_z = FALSE,
+)
+	var/static/parsed_maps = list()
+	if(!(dmm_file in parsed_maps))
+		parsed_maps[dmm_file] = new /datum/parsed_map(dmm_file)
+
+	var/datum/parsed_map/parsed_map = parsed_maps[dmm_file]
```
don't. not safe. We modify the parsed map datum mutates itself as we load, it isn't safe to reuse it.

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -117,20 +117,59 @@
+ * - place_on_top: Whether to use /turf/proc/PlaceOnTop rather than /turf/proc/ChangeTurf
+ * - new_z: If true, a new z level will be created for the map
+ */
+/proc/load_map(
+	dmm_file,
+	x_offset = 0,
+	y_offset = 0,
+	z_offset = 0,
+	only_load_this_z = 0,
+	crop_map_to_world = FALSE,
+	measure_only = FALSE,
+	no_changeturf = FALSE,
+	x_lower = -INFINITY,
+	x_upper = INFINITY,
+	y_lower = -INFINITY,
+	y_upper = INFINITY,
+	place_on_top = FALSE,
+	new_z = FALSE,
+)
```
fuck you fuck you fuck you fuck you fuck you why are you reformatting my procs

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -117,20 +117,59 @@
+	x_offset = 0,
+	y_offset = 0,
+	z_offset = 0,
+	only_load_this_z = 0,
```
you did not document this. also why does it exist 
implement z_lower and z_upper (challenge, do it optimally in the same way we handle x and y, may not be possible with tgm we'll see)

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -243,11 +282,37 @@
-/datum/parsed_map/proc/load(x_offset, y_offset, z_offset, cropMap, no_changeturf, x_lower, x_upper, y_lower, y_upper, placeOnTop, whitelist = FALSE, new_z)
+/// Load the parsed map into the world. You probably want [/proc/load_map]. Keep the signature the same.
+/datum/parsed_map/proc/load(
+	x_offset = 0,
+	y_offset = 0,
+	z_offset = 0,
+	only_load_this_z = 0,
+	crop_map_to_world = FALSE,
+	no_changeturf = FALSE,
+	x_lower = -INFINITY,
+	x_upper = INFINITY,
+	y_lower = -INFINITY,
+	y_upper = INFINITY,
+	place_on_top = FALSE,
+	new_z = FALSE,
+)
 	//How I wish for RAII
 	Master.StartLoadingMap()
-	. = _load_impl(x_offset, y_offset, z_offset, cropMap, no_changeturf, x_lower, x_upper, y_lower, y_upper, placeOnTop, new_z)
+	. = _load_impl(
+		x_offset,
+		y_offset,
+		z_offset,
+		only_load_this_z,
+		crop_map_to_world,
+		no_changeturf,
+		x_lower,
+		x_upper,
+		y_lower,
+		y_upper,
+		place_on_top,
+		new_z,
```
I am killing you

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -378,22 +496,25 @@
+	var/highest_z = only_load_this_z ? 1 : last_column.zcrd + z_offset - 1 // Lets not just make a new z level each time we increment maxz
 	var/z_threshold = world.maxz
-	if(highest_z > z_threshold && cropMap)
+	if(highest_z > z_threshold && crop_map_to_world)
```
just call it crop map

LemonInTheDark commented on code/modules/mapping/space_management/multiz_helpers.dm:
```diff
@@ -28,24 +29,30 @@
 		return (dir | get_dir(us, them))
 
 /turf/proc/above()
+	if(turf_flags & RESERVATION_TURF)
+		var/datum/turf_reservation/map_reserve = SSmapping.get_reservation_from_turf(src)
+		return map_reserve.get_turf_above(src)
 	return GET_TURF_ABOVE(src)
 
 /turf/proc/below()
+	if(turf_flags & RESERVATION_TURF)
+		var/datum/turf_reservation/map_reserve = SSmapping.get_reservation_from_turf(src)
+		return map_reserve.get_turf_below(src)
 	return GET_TURF_BELOW(src)
```
bro this proc is a helper, add this to the macro don't just revert all uses of it. It's like that because it's very hot in multiz and I want to avoid proc overhead

LemonInTheDark reviewed (changes requested):
mostly parsing stuff (I am upset because of bullshit microsoft shit sorry if I'm harsh)

ZephyrTFA committed:
work

ZephyrTFA committed:
naming

ZephyrTFA committed:
l

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -114,23 +114,78 @@
+    newfriend.gridSets = gridSets.Copy()
+    newfriend.modelCache = modelCache.Copy()
+    newfriend.parsed_bounds = parsed_bounds.Copy()
+    newfriend.bounds = parsed_bounds.Copy()
```
```suggestion
    // Copy parsed bounds to reset to initial values
    newfriend.bounds = parsed_bounds.Copy()
```

ZephyrTFA committed:
you can be null, code comment

ZephyrTFA committed:
reversiom

ZephyrTFA committed:
adjust macros

ZephyrTFA committed:
spaced

ZephyrTFA committed:
g

ZephyrTFA committed:
Merge branch 'multiztemplates' of github.com:ZephyrTFA/tgstation into multiztemplates

ZephyrTFA committed:
cordon turfs aren't technically part of a reservation

ZephyrTFA commented:
🕛

Ghommie commented on code/modules/mapping/reader.dm:
```diff
@@ -114,23 +114,67 @@
+	y_offset = 0,
+	z_offset = 0,
+	only_load_this_z = 0,
+	crow_map = FALSE,
```
shouldn't this arg be called `crop_map`? This looks like a typo.

ZephyrTFA replied:
as straight as the path a crow flies

EuSouAFazer commented:
Yooooo. Massive Multi-Z W, you're doing god's work here. Massive W

ZephyrTFA commented on code/modules/mapping/reader.dm:
```diff
@@ -114,23 +114,67 @@
+ * - y_offset: The y offset to load the map at
+ * - z_offset: The z offset to load the map at
+ * - only_load_this_z: If set, only this z level from the given map will be loaded
+ * - crow_map: If true, the map will be cropped to the world bounds
+ * - measure_only: If true, the map will not be loaded, but the bounds will be calculated
+ * - no_changeturf: If true, the map will not call /turf/AfterChange
+ * - x_lower: The minimum x coordinate to load
+ * - x_upper: The maximum x coordinate to load
+ * - y_lower: The minimum y coordinate to load
+ * - y_upper: The maximum y coordinate to load
+ * - place_on_top: Whether to use /turf/proc/PlaceOnTop rather than /turf/proc/ChangeTurf
+ * - new_z: If true, a new z level will be created for the map
+ */
+/proc/load_map(
+	dmm_file,
+	x_offset = 0,
+	y_offset = 0,
+	z_offset = 0,
+	only_load_this_z = 0,
+	crow_map = FALSE,
+	measure_only = FALSE,
+	no_changeturf = FALSE,
+	x_lower = -INFINITY,
+	x_upper = INFINITY,
+	y_lower = -INFINITY,
+	y_upper = INFINITY,
+	place_on_top = FALSE,
+	new_z = FALSE,
+)
+	if(!(dmm_file in GLOB.cached_maps))
+		GLOB.cached_maps[dmm_file] = new /datum/parsed_map(dmm_file)
+
+	var/datum/parsed_map/parsed_map = GLOB.cached_maps[dmm_file]
+	parsed_map = parsed_map.copy()
+	if(!measure_only && !isnull(parsed_map.bounds))
+		parsed_map.load(x_offset, y_offset, z_offset, only_load_this_z, crow_map, no_changeturf, x_lower, x_upper, y_lower, y_upper, place_on_top, new_z)
```
```suggestion
 * - crop_map: If true, the map will be cropped to the world bounds
 * - measure_only: If true, the map will not be loaded, but the bounds will be calculated
 * - no_changeturf: If true, the map will not call /turf/AfterChange
 * - x_lower: The minimum x coordinate to load
 * - x_upper: The maximum x coordinate to load
 * - y_lower: The minimum y coordinate to load
 * - y_upper: The maximum y coordinate to load
 * - place_on_top: Whether to use /turf/proc/PlaceOnTop rather than /turf/proc/ChangeTurf
 * - new_z: If true, a new z level will be created for the map
 */
/proc/load_map(
	dmm_file,
	x_offset = 0,
	y_offset = 0,
	z_offset = 0,
	only_load_this_z = 0,
	crop_map = FALSE,
	measure_only = FALSE,
	no_changeturf = FALSE,
	x_lower = -INFINITY,
	x_upper = INFINITY,
	y_lower = -INFINITY,
	y_upper = INFINITY,
	place_on_top = FALSE,
	new_z = FALSE,
)
	if(!(dmm_file in GLOB.cached_maps))
		GLOB.cached_maps[dmm_file] = new /datum/parsed_map(dmm_file)

	var/datum/parsed_map/parsed_map = GLOB.cached_maps[dmm_file]
	parsed_map = parsed_map.copy()
	if(!measure_only && !isnull(parsed_map.bounds))
		parsed_map.load(x_offset, y_offset, z_offset, only_load_this_z, crop_map, no_changeturf, x_lower, x_upper, y_lower, y_upper, place_on_top, new_z)
```

ZephyrTFA committed:
Update code/modules/mapping/reader.dm

jlsnow301 committed:
change to upper/lower

jlsnow301 committed:
Merge branch 'master' into multiztemplates

ZephyrTFA committed:
Merge branch 'multiztemplates' of github.com:ZephyrTFA/tgstation into multiztemplates

ZephyrTFA commented:
@LemonInTheDark 📆

LemonInTheDark commented on code/__DEFINES/multiz.dm:
```diff
@@ -1,4 +1,8 @@
 /// Attempt to get the turf below the provided one according to Z traits
-#define GET_TURF_BELOW(turf) ((!(turf) || !length(SSmapping.multiz_levels) || !SSmapping.multiz_levels[(turf).z][Z_LEVEL_DOWN]) ? null : get_step((turf), DOWN))
+#define GET_TURF_BELOW(turf) ( \
+	(turf:turf_flags & RESERVATION_TURF) ? SSmapping.get_reservation_from_turf(turf)?.get_turf_below(turf) : \
```
is : actually needed here? I'm fine with forcing people to properly type the arg to this macro

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -114,26 +114,75 @@
+	place_on_top = FALSE,
+	new_z = FALSE,
+)
+	if(z_lower < 1)
+		CRASH("Attempted to call load_map with a z_lower less than 1, this will not work")
```
why? not the case for the other two types

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -114,26 +114,75 @@
+ * - place_on_top: Whether to use /turf/proc/PlaceOnTop rather than /turf/proc/ChangeTurf
+ * - new_z: If true, a new z level will be created for the map
+ */
+/proc/load_map(
+	dmm_file,
+	x_offset = 0,
+	y_offset = 0,
+	z_offset = 0,
+	crop_map = FALSE,
+	measure_only = FALSE,
+	no_changeturf = FALSE,
+	x_lower = -INFINITY,
+	x_upper = INFINITY,
+	y_lower = -INFINITY,
+	y_upper = INFINITY,
+	z_lower = 1,
+	z_upper = INFINITY,
+	place_on_top = FALSE,
+	new_z = FALSE,
+)
```
still not jazzed about this but I do get it

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -330,10 +387,10 @@
 		if(new_z)
 			// Need to avoid improperly loaded area/turf_contents
-			world.increaseMaxY(highest_y, max_zs_to_load = z_offset - 1)
+			world.increaseMaxY(highest_y, max_zs_to_load = z_offset - z_lower)
```
is this right? why would loading in 1 less z layer mean I don't want to increase maxy all the way
z_offset is the lowest z level right?

ZephyrTFA replied:
no idea, just copying previous behaviour. (I'm fairly certain the value isnt even used for modifying maxx and maxy

ZephyrTFA replied:
figured it out

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -367,33 +423,36 @@
 		if(new_z)
 			// Need to avoid improperly loaded area/turf_contents
-			world.increaseMaxX(final_x, max_zs_to_load = z_offset - 1)
+			world.increaseMaxX(final_x, max_zs_to_load = z_offset - z_lower)
```
similar, applies to dmm too I expect

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -367,33 +423,36 @@
 		if(!no_changeturf)
 			WARNING("Z-level expansion occurred without no_changeturf set, this may cause problems when /turf/AfterChange is called")
 
 	for(var/datum/grid_set/gset as anything in gridSets)
+		if(gset.zcrd > z_upper || gset.zcrd < z_lower)
+			continue // skip z levels we dont want
```
since limiting z layer loads is an uncommon case, I think it would be better to conditionally filter gridsets if z_upper/z_lower are set to any interesting values.

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -367,33 +423,36 @@
 			continue
 
-		var/zcrd = gset.zcrd + z_offset - 1
+		var/zcrd = gset.zcrd + z_offset - z_lower
```
if we're gonna do this op a lot (z offset - z lower), you should do it as a var at the start since it won't change as we go (I assume)
saves cpu, and cleans code potentially if you pick a good name

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -456,22 +515,25 @@
 	var/x_relative_to_absolute = x_offset - 1
 	var/line_len = src.line_len
 	for(var/datum/grid_set/gset as anything in gridSets)
+		if(gset.zcrd > z_upper || gset.zcrd < z_lower)
+			continue // skip z levels we dont want
+
```
don't need to micro this since it's not a case used to load on live

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -456,22 +515,25 @@
 		var/ycrd = relative_y + y_relative_to_absolute
-		var/zcrd = gset.zcrd + z_offset - 1
-		if(!cropMap && ycrd > world.maxy)
+		var/zcrd = gset.zcrd + z_offset - z_lower
```
still might want a var for the z_offset - z_lower thing tho

ZephyrTFA committed:
Merge remote-tracking branch 'upstream/master' into multiztemplates

ZephyrTFA committed:
change macro to use qualified access instead of unqualified

ZephyrTFA committed:
z filtering is done only when bounds are set
change name of handling procs for maxx and maxy
change parameter names to be more descriptive
also add codedoc for them

norsvenska commented:
multi-z shuttles when

LemonInTheDark committed:
Fixes infinite lower_z value

ZephyrTFA committed:
that was bad; thank god for unit tests

ZephyrTFA committed:
better z upper offset parsing

ZephyrTFA committed:
template signal for loaded

ZephyrTFA committed:
VICTORY

ZephyrTFA committed:
dont forget dmm

ZephyrTFA committed:
Merge remote-tracking branch 'upstream/master' into multiztemplates

ZephyrTFA committed:
irritate

LemonInTheDark commented on code/datums/lazy_template.dm:
```diff
@@ -46,25 +44,55 @@
+			for(var/thing in turf.get_all_contents())
+				to_init |= thing
+
+		SSatoms.InitializeAtoms(to_init)
+		for(var/loaded_atom in to_init)
+			my_loaded_atoms |= loaded_atom
```
why are you using |= here. why add at all why not just set my_loaded_atoms to match to_init how would dupes get in here

LemonInTheDark commented on code/datums/lazy_template.dm:
```diff
@@ -46,25 +44,55 @@
+			z_lower = z_idx,
+		)
+		var/list/to_init = list()
+		for(var/turf/turf as anything in block(bottom_left, top_right))
+			to_init |= get_area(turf)
+			to_init |= turf
+			for(var/thing in turf.get_all_contents())
+				to_init |= thing
```
at the very least turfs will never duplicate, why use |= for them.
frankly this would be faster if you had a list for areas, a list for turfs and a list for objects, and then +='d them together when you finished.

LemonInTheDark commented on code/datums/lazy_template.dm:
```diff
@@ -46,25 +44,55 @@
+			my_loaded_atoms |= loaded_atom
+
+	SEND_SIGNAL(src, COMSIG_LAZY_TEMPLATE_LOADED, my_loaded_atoms)
+	my_loaded_atoms.Cut() // We don't want the list to cause ref holding shenanigans due to ASYNC
```
nothing between this line and the var going out of scope sleeps, so this should do nothing. what are you trying to guard against?

ZephyrTFA replied:
people saving references to the list by using a signal handler that just does local_list = my_loaded_atoms

LemonInTheDark replied:
isn't that a them problem? they could also copy it and cause the same problem. what's the point of trying to block against it like this

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -367,33 +431,51 @@
 	var/lowest_x = max(x_lower, 1 - x_relative_to_absolute)
 
+	// Amount we offset the grid zcrd to get the true zcrd
+	var/grid_z_offset = z_offset - 1
+	var/z_upper_set = z_upper < INFINITY
+	var/z_lower_set = z_lower > -INFINITY
+
 	// We make the assumption that the last block of turfs will have the highest embedded z in it
-	var/highest_z = last_column.zcrd + z_offset - 1 // Lets not just make a new z level each time we increment maxz
+	// true max zcrd
+	var/map_bounds_z_max = last_column.zcrd
+	var/z_upper_parsed = map_bounds_z_max + z_offset - 1
+	if(z_upper_set)
+		z_upper_parsed -= map_bounds_z_max - z_upper
+	if(z_lower_set)
+		var/offset_amount = z_lower - 1
+		z_upper_parsed -= offset_amount
+		grid_z_offset -= offset_amount
+
+	var/list/target_grid_sets = gridSets
+	if(z_lower_set || z_upper_set) // bounds are set, filter out gridsets for z levels we don't want
+		target_grid_sets = filter_grid_sets_based_on_z_bounds(z_lower, z_upper)
```
was this new system tested?

ZephyrTFA replied:
yes it was, it's why it took me a little bit to get it pushed as I was testing it and fixing bugs

LemonInTheDark commented on code/modules/mapping/reader.dm:
```diff
@@ -367,33 +431,51 @@
 			WARNING("Z-level expansion occurred without no_changeturf set, this may cause problems when /turf/AfterChange is called")
 
-	for(var/datum/grid_set/gset as anything in gridSets)
+	for(var/datum/grid_set/gset as anything in target_grid_sets)
```
this name makes me depressed

ZephyrTFA replied:
target_grid_sets or gset;

If gset I blame you but can also make it better

LemonInTheDark replied:
I can live with this I guess

ZephyrTFA committed:
Merge remote-tracking branch 'upstream/master' into multiztemplates

ZephyrTFA committed:
>:(

ZephyrTFA committed:
split atom tracking for lazy templates into two lists

LemonInTheDark commented on code/datums/lazy_template.dm:
```diff
@@ -74,15 +74,23 @@
+
+		// turfs can never be duplicated so keep them out of the below list to optimize O(n)
+		var/list/turfs = list()
+		// everything else that can be duplicated, such as areas or large objects
+		var/list/not_turfs = list()
```
I mentioned areas because it'd be faster to keep them separate, since |= is O(N) and if we only need to do it a few times for each area that'll save time.
so basically do this but have a list for areas and one for movables

ZephyrTFA committed:
split up areas too

ZephyrTFA committed:
dont cut the list, trust the frontend coder
